### ISSUE TITLE
Win32::GetLastError fails when first called

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8896]
Migrated from [rt.perl.org#42925](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=42925) (status was 'resolved')

Searchable as RT42925$

**Comment by p5pRT** [comment]
### From nobull@cpan.org

This is a bug report for perl from nobull@&#8203;cpan\.org\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.




---
The first time Win32&#8203;::GetLastError is called it always returns 2 \("The
system cannot find the file specified"\)\. \(This in turn this breaks
Win32&#8203;::TieRegistry etc but the underlying problem is very simple to
illustrate\)\.

Note&#8203;: Win32&#8203;::GetLastError is \_core\_ it's not part of the Win32 module\, and
indeed if you actually load Win32\.pm you will \_not\_ see this error manifest\.

\#\!perl
use strict;
use warnings;
\#Win32&#8203;::GetLastError\(\); \# Uncomment to see correct error below\.
kill 0\,1; \# Cause Windows error 6 "The handle is invalid"
my $e = Win32&#8203;::GetLastError\(\);
print $e\," "\,Win32&#8203;::FormatMessage\($e\)\,"\\n";
\_\_END\_\_



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.8&#8203;:

Configured by SYSTEM at Tue Jan 23 15&#8203;:57&#8203;:26 2007\.

Summary of my perl5 \(revision 5 version 8 subversion 8\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags ='\-nologo \-GF \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32
\-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT \-DNO\_HASH\_SEED

\-DUSE\_SITECUSTOMIZE \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO
\-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion='12\.00\.8804'\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf
\-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth=\\lib
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib

oleaut32\.lib  netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib
version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib

oleaut32\.lib  netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib
version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl58\.lib
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug
\-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;    ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
&nbsp;    Iin\_load\_module moved for compatibility with build 806
&nbsp;    Avoid signal flag SA\_RESTART for older versions of HP\-UX
&nbsp;    PerlEx support in CGI&#8203;::Carp
&nbsp;    Less verbose ExtUtils&#8203;::Install and Pod&#8203;::Find
&nbsp;    Patch for CAN\-2005\-0448 from Debian with modifications
&nbsp;    Rearrange @&#8203;INC so that 'site' is searched before 'perl'
&nbsp;    Partly reverted 24733 to preserve binary compatibility
&nbsp;    29930 win32\.c typo in \#define MULTIPLICITY
&nbsp;    29868 win32\_async\_check\(\) can still loop indefinitely
&nbsp;    29690\,29732 ANSIfy the PATH environment variable on Windows
&nbsp;    29689 Add error handling to win32\_ansipath
&nbsp;    29675 Use short pathnames in $^X and @&#8203;INC
&nbsp;    29607\,29676 allow blib\.pm to be used for testing Win32 module
&nbsp;    29605 Implement killpg\(\) for MSWin32
&nbsp;    29598 cwd\(\) to return the short pathname
&nbsp;    29597 let readdir\(\) return the alternate filename
&nbsp;    29590 Don't destroy the Unicode system environment on Perl startup
&nbsp;    29528 get ext/Win32/Win32\.xs to compile on cygwin
&nbsp;    29509\,29510\,29511 Move Win32&#8203;::\* functions into Win32 module
&nbsp;    29483 Move Win32 from win32/ext/Win32 to ext/Win32
&nbsp;    29481 Makefile\.PL changes to compile Win32\.xs using cygwin
&nbsp;    28671 Define PERL\_NO\_DEV\_RANDOM on Windows
&nbsp;    28376 Add error checks after execing PL\_cshname or PL\_sh\_path
&nbsp;    28305 Pod&#8203;::Html should not convert \\"foo\\" into \`\`foo''
&nbsp;    27833 Change anchor generation in Pod&#8203;::Html for '=item item 2'
&nbsp;    27832\,27847 fix Pod&#8203;::Html&#8203;::depod\(\) for multi\-line strings
&nbsp;    27736 Make perl\_fini\(\) run with Sun WorkShop compiler
&nbsp;    27719 Document the functions htmlify\(\) and anchorify\(\) in Pod&#8203;::Html
&nbsp;    27619 Bug in Term&#8203;::ReadKey being triggered by a bug in Term&#8203;::ReadLine
&nbsp;    27549 Move DynaLoader\.o into libperl\.so
&nbsp;    27528 win32\_pclose\(\) error exit doesn't unlock mutex
&nbsp;    27527 win32\_async\_check\(\) can loop indefinitely
&nbsp;    27515 ignore directories when searching @&#8203;INC
&nbsp;    27359 Fix \-d&#8203;:Foo=bar syntax
&nbsp;    27210 Fix quote typo in c2ph
&nbsp;    27203 Allow compiling swigged C\+\+ code
&nbsp;    27200 Make stat\(\) on Windows handle trailing slashes correctly
&nbsp;    27194 Get perl\_fini\(\) running on HP\-UX again
&nbsp;    27133 Initialise lastparen in the regexp structure
&nbsp;    27061 L\<PerlIO> and Pod&#8203;::Html
&nbsp;    27034 Avoid \\"Prototype mismatch\\" warnings with autouse
&nbsp;    26970 Make Passive mode the default for Net&#8203;::FTP
&nbsp;    26921 Avoid getprotobyname/number calls in IO&#8203;::Socket&#8203;::INET
&nbsp;    26897\,26903 Make common IPPROTO\_\* constants always available
&nbsp;    26670 Make '\-s' on the shebang line parse \-foo=bar switches
&nbsp;    26637 Make Borland and MinGW happy with change 26379
&nbsp;    26536 INSTALLSCRIPT versus INSTALLDIRS
&nbsp;    26379 Fix alarm\(\) for Windows 2003
&nbsp;    26087 Storable 0\.1 compatibility
&nbsp;    25861 IO&#8203;::File performace issue
&nbsp;    25084 long groups entry could cause memory exhaustion
&nbsp;    24699 ICMP\_UNREACHABLE handling in Net&#8203;::Ping



---
@&#8203;INC for perl v5\.8\.8&#8203;:
&nbsp;    C&#8203;:/Perl/site/lib
&nbsp;    C&#8203;:/Perl/lib
&nbsp;    \.



---
Environment for perl v5\.8\.8&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=C&#8203;:\\Program Files\\Windows Resource

Kits\\Tools\\;C&#8203;:\\PROGRA~1\\Borland\\Delphi5\\Projects\\Bpl;C&#8203;:\\PROGRA~1\\Borland\\Delphi5\\Bin;C&#8203;:\\Perl\\site\\bin;C&#8203;:\\Perl\\bin;C&#8203;:\\WINDOWS\\

system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;;C&#8203;:\\MSSQL\\BINN;C&#8203;:\\Program
Files\\Common Files\\Adaptec Shared\\System;C&#8203;:\\Program

Files\\Gemplus\\GemSafe Libraries User\\BIN;C&#8203;:\\PROGRA~1\\Gemplus\\GAC;C&#8203;:\\Program
Files\\Support Tools\\;C&#8203;:\\Program Files\\Microsoft

SQL Server\\80\\Tools\\Binn\\;C&#8203;:\\Program Files\\Microsoft SQL
Server\\90\\Tools\\binn\\;C&#8203;:\\Program Files\\Microsoft SQL

Server\\90\\DTS\\Binn\\;C&#8203;:\\Program Files\\Microsoft SQL
Server\\90\\Tools\\Binn\\VSShell\\Common7\\IDE\\;C&#8203;:\\Program Files\\Microsoft

Visual Studio 8\\Common7\\IDE\\PrivateAssemblies\\;R&#8203;:\\bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @jandubois

On Fri\, 11 May 2007\, Brian McCauley \(via RT\) wrote&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> The first time Win32&#8203;::GetLastError is called it always returns 2 \("The
> system cannot find the file specified"\)\. \(This in turn this breaks
> Win32&#8203;::TieRegistry etc but the underlying problem is very simple to
> illustrate\)\.
>
> Note&#8203;: Win32&#8203;::GetLastError is \_core\_ it's not part of the Win32 module\, and
> indeed if you actually load Win32\.pm you will \_not\_ see this error manifest\.

Thank you for the bug report\!  All the Win32&#8203;::\* functions have been moved
to the Win32 module\, and the core only contains forwarding stubs for those
functions that traditionally did not require a "use Win32" statement\.

This forwarding stub doesn't preserve the value of the last error\, which
is the reason that you see the wrong value when the code goes through the
forwarder\.  The attached patch fixes that\.
 
> \#\!perl
> use strict;
> use warnings;
> \#Win32&#8203;::GetLastError\(\); \# Uncomment to see correct error below\.

I disagree with the comment above&#8203;: There is nothing in the documentation
of the kill\(\) function that promises specific values in Win32&#8203;::GetLastError\(\)
when the function fails\.  Indeed you'll see the following error with the
latest implementation changes to the kill function&#8203;:

87 The parameter is incorrect\.

But that is just a side\-note; Win32&#8203;::GetLastError\(\) should of course return
consistent results even when called via the forwarder\.

> kill 0\,1; \# Cause Windows error 6 "The handle is invalid"
> my $e = Win32&#8203;::GetLastError\(\);
> print $e\," "\,Win32&#8203;::FormatMessage\($e\)\,"\\n";
> \_\_END\_\_

Cheers\,
\-Jan

<details><summary>Inline Patch</summary>

```diff
--- ext/Win32CORE/Win32CORE.c.orig	Thu Feb 22 02:12:25 2007
+++ ext/Win32CORE/Win32CORE.c	Fri May 11 16:45:27 2007
@@ -15,7 +15,9 @@
 forward(pTHX_ const char *function)
 {
     dXSARGS;
+    DWORD err = GetLastError();
     Perl_load_module(aTHX_ PERL_LOADMOD_NOIMPORT, newSVpvn("Win32",5), newSVnv(0.27));
+    SetLastError(err);
     SPAGAIN;
     PUSHMARK(SP-items);
     call_pv(function, GIMME_V);
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @steve-m-hay

Jan Dubois wrote&#8203;:
> On Fri\, 11 May 2007\, Brian McCauley \(via RT\) wrote&#8203;:
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> The first time Win32&#8203;::GetLastError is called it always returns 2 \("The
>> system cannot find the file specified"\)\. \(This in turn this breaks
>> Win32&#8203;::TieRegistry etc but the underlying problem is very simple to
>> illustrate\)\.
>>
>> Note&#8203;: Win32&#8203;::GetLastError is \_core\_ it's not part of the Win32 module\, and
>> indeed if you actually load Win32\.pm you will \_not\_ see this error manifest\.
> 
> Thank you for the bug report\!  All the Win32&#8203;::\* functions have been moved
> to the Win32 module\, and the core only contains forwarding stubs for those
> functions that traditionally did not require a "use Win32" statement\.
> 
> This forwarding stub doesn't preserve the value of the last error\, which
> is the reason that you see the wrong value when the code goes through the
> forwarder\.  The attached patch fixes that\.

Thanks\, applied as \#31214 \(with a test also applied as \#31215\)\.

\-\-

**Comment by p5pRT** [comment]
### From @jandubois

On Mon\, 14 May 2007\, Steve Hay wrote&#8203;:
> Jan Dubois wrote&#8203;:
>> This forwarding stub doesn't preserve the value of the last error\,
>> which is the reason that you see the wrong value when the code goes
>> through the forwarder\. The attached patch fixes that\.
>
> Thanks\, applied as \#31214 \(with a test also applied as \#31215\)\.

I would make sure that Win32 isn't already loaded when you call
Win32&#8203;::GetLastError\(\)\.  Suggested patch attached\.

Cheers\,
\-Jan
 
<details><summary>Inline Patch</summary>

```diff
--- ext/Win32CORE/t/win32core.t.orig	Mon May 14 09:58:14 2007
+++ ext/Win32CORE/t/win32core.t	Mon May 14 15:10:10 2007
@@ -10,11 +10,17 @@
 	}
     }
 
-    plan tests => 2;
+    plan tests => 4;
 };
 use_ok( "Win32CORE" );
 
+# Make sure the Win32 is not yet loaded
+ok(!defined &Win32::ExpandEnvironmentStrings);
+
 # [perl #42925] - Loading Win32::GetLastError() via the forwarder function
 # should not affect the last error being retrieved
 $^E = 42;
 is(Win32::GetLastError(), 42, 'GetLastError() works on the first call');
+
+# Now all Win32::* functions should be loaded
+ok(defined &Win32::ExpandEnvironmentStrings);
End of Patch
```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

Jan Dubois wrote&#8203;:
> On Mon\, 14 May 2007\, Steve Hay wrote&#8203;:
>> Jan Dubois wrote&#8203;:
>>> This forwarding stub doesn't preserve the value of the last error\,
>>> which is the reason that you see the wrong value when the code goes
>>> through the forwarder\. The attached patch fixes that\.
>> Thanks\, applied as \#31214 \(with a test also applied as \#31215\)\.
> 
> I would make sure that Win32 isn't already loaded when you call
> Win32&#8203;::GetLastError\(\)\.  Suggested patch attached\.

Thanks\, applied as \#31219\.

\-\-

**Comment by p5pRT** [comment]
### From @steve-m-hay

Now resolved in bleadperl by changes \#31214\, 31215 and 31219\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
